### ISSUE TITLE
Resolving special variables can fail with confusing message

### ISSUE THREAD

**Issue by bhirsz** [robotframework/robotframework#4898]
Considering such example:

```
*** Variables ***
${TABLE}    example


*** Test Cases ***
Special Variables
    IF    all(item not in $TABLE for item in ("one", "two"))
        Log    ${TABLE}
    ELSE
        Log    not ${TABLE}
    END

```

I'm not sure if ``all(item not in $TABLE for item in ("one", "two"))`` is supported at all but the error I'm getting looks incorrect - it contains 'internal' naming used in RF (with prefix ``RF_VAR_``):

```
NameError: name 'RF_VAR_TABLE' is not defined
```

**Comment by pekkaklarck** [comment]
The issue is due to Python's variable scoping rules and how we handle variables using the `$name` syntax. These special variables are handled so that we [first convert the `$` to `RF_VAR_` prefix](https://github.com/robotframework/robotframework/blob/f8b7b88214737100f66cff5a9001b0c53607fb1b/src/robot/variables/evaluation.py#L70) and then when the [expression is evaluated](https://github.com/robotframework/robotframework/blob/f8b7b88214737100f66cff5a9001b0c53607fb1b/src/robot/variables/evaluation.py#L67) we pass a [custom mapping](https://github.com/robotframework/robotframework/blob/f8b7b88214737100f66cff5a9001b0c53607fb1b/src/robot/variables/evaluation.py#L126) as a local namespace and that mapping handles all variables starting with `RF_VAR`.

Why the above doesn't work in this case is that the comprehension creates a new scope and local variables we pass to `eval()` aren't seen there. Global variables would be seen, but for some reason `eva()` requires that globals are a real dict, not a custom mapping like we have.

One possible way to avoid this problem would be creating a real dict that contains all Robot's variables with the `RF_VAR` prefix and use that as globals. That would make the `$name` syntax case-sensitive, though, which would be a backwards incompatible change. It's also possible that creating the dict would have a measurable performance penalty if expressions are evaluated often. We'd also anyway need our current custom mapping, because it also handles importing modules used in expressions.

Unfortunately the possible fix doesn't sound too tempting and I don't have any better ideas how to fix this. We could, however, catch the `NameError` and reraise it with some more information about the problem if the message contains `RF_VAR_`.

**Comment by pekkaklarck** [comment]
Here's a version of the expression that works:
```
IF    (lambda table=$TABLE: all(item not in table for item in ("one", "two")))()
```
The reason it works is that the `table` argument is available in the lambda function scope and such intermediate scopes are seen by comprehensions. `$TABLE` itself wouldn't be seen, but it was set to be the default value of `table` in a higher level scope where it's still seen. This is rather horrible and I don't recommend using it as a workaround. :smiley:

**Comment by bhirsz** [comment]
That's what I suspected as well (I'm bit of familiar with the implementation of $var handling) and I know it may be difficult to solve that. But what about raising more detailed error message? Like you suggested, reraising NameError - most users will be suprised by ``RF_VAR_TABLE``. Something along the lines of 'unrecognized variable $table or variable cannot be resolved' would be more readable.

**Comment by pekkaklarck** [comment]
Yeah, let's try to enhance the error. We can easily catch the NameError, and I guess we can just check does the error contain `RF_VAR_`. If contains, we can raise a new NameError with more informative message. Are you @bhirsz interested to look at this? I'll add this to RF 7.0 scope nevertheless.

**Comment by pekkaklarck** [comment]
I fixed this so that the original example now fails like this:
> Invalid IF condition: Evaluating expression 'all(item not in $TABLE for item in ("one", "two"))' failed: Robot Framework variable '$TABLE' used in the expression part of a comprehension or some other scope where it cannot be seen.

**Comment by pekkaklarck** [comment]
It seems that evaluation rules have changed in Python 3.12. The test I added for the enhanced error message fails because the test actually passes and the following experiment proves that:

    $ python3.12 -c 'print(eval("[x + i for i in (0, 1, 2)]", {}, {"x": 1}))'
    [1, 2, 3]
    $ python3.11 -c 'print(eval("[x + i for i in (0, 1, 2)]", {}, {"x": 1}))'
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "<string>", line 1, in <module>
      File "<string>", line 1, in <listcomp>
    NameError: name 'x' is not defined

Can you @bhirsz test Python 3.12 in your environment as well?

**Comment by pekkaklarck** [comment]
The change seems to be due to [PEP-709](https://peps.python.org/pep-0709/).

**Comment by bhirsz** [comment]
I confirm that Python 3.12 behaves differently. Hovewer the fix is still working:

Robot 6.1.1 before fix:

![image](https://github.com/robotframework/robotframework/assets/8532066/b93f225f-af33-4478-97b4-3e08e8a198da)

 NameError: name 'RF_VAR_TABLE' is not defined

After fix (7.0.dev1):

![image](https://github.com/robotframework/robotframework/assets/8532066/9068a816-39eb-4413-bc69-93748f974600)

Robot Framework variable '$TABLE' is used in a scope where it cannot be seen.

**Comment by pekkaklarck** [comment]
I was surprised that the test failed, but it seems that PEP-709 affected only comprehensions, not generator expressions:
```python
>>> eval("[i + x for i in (0, 1, 2)]", {}, {'x': 1})
[1, 2, 3]
>>> eval("{i + x for i in (0, 1, 2)}", {}, {'x': 1})
{1, 2, 3}
>>> eval("{i: x for i in (0, 1, 2)}", {}, {'x': 1})
{0: 1, 1: 1, 2: 1}
>>> eval("list(i + x for i in (0, 1, 2))", {}, {'x': 1})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
  File "<string>", line 1, in <genexpr>
NameError: name 'x' is not defined
```
This means that using
```
IF    all([item not in $TABLE for item in ("one", "two")])
```
ought to work. It's a lot better workaround than the lambda hack I presented earlier.
